File: scripts/check-licenses.mjs
#!/usr/bin/env node
// Гейт лицензий npm-зависимостей (B52 — контроль copyleft).
//
// Читает frontend/package-lock.json и завершается с ошибкой (exit 1), если
// среди зависимостей встречается strong/weak copyleft, несовместимый с
// проприетарной поставкой (GPL / AGPL / LGPL / SSPL), либо у prod-зависимости
// лицензия не определена (UNKNOWN). Namespace npm LGPL держим запрещённым:
// на этой стороне его нет (в отличие от Python-драйвера psycopg, см.
// scripts/check_licenses.py и reports/B52-licenses.md).
//
// Без внешних зависимостей — разбираем lock-файл напрямую, чтобы гейт сам не
// тянул новых пакетов (та же логика, что применялась при аудите).

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

const here = dirname(fileURLToPath(import.meta.url));
const lockPath = resolve(here, '..', 'frontend', 'package-lock.json');

/** Достаёт строку лицензии из записи lock-файла (string | {type} | licenses[]). */
function licenseString(entry) {
  if (!entry) return '';
  if (typeof entry.license === 'string') return entry.license;
  if (entry.license && typeof entry.license === 'object' && entry.license.type) {
    return entry.license.type;
  }
  if (Array.isArray(entry.licenses)) {
    return entry.licenses.map((l) => (typeof l === 'string' ? l : l.type || '')).join(' ');
  }
  return '';
}

/** Возвращает вид запрещённой лицензии или null. Порядок важен: AGPL/LGPL до GPL. */
function forbiddenLicense(lic) {
  const s = String(lic).toUpperCase();
  if (/AGPL/.test(s)) return 'AGPL';
  if (/LGPL/.test(s)) return 'LGPL';
  if (/(^|[^L])GPL/.test(s)) return 'GPL';
  if (/SSPL/.test(s)) return 'SSPL';
  return null;
}

const lock = JSON.parse(readFileSync(lockPath, 'utf8'));
const packages = lock.packages || {};

const violations = [];
const unknownProd = [];

for (const [path, entry] of Object.entries(packages)) {
  if (path === '') continue; // корневой пакет — наш собственный код
  if (!path.includes('node_modules/')) continue;
  const name = path.replace(/.*node_modules\//, '');
  const lic = licenseString(entry);
  const bad = forbiddenLicense(lic);
  if (bad) {
    violations.push({ name, version: entry.version, license: lic || '(нет)', kind: bad, dev: !!entry.dev });
  } else if (!lic && !entry.dev) {
    unknownProd.push({ name, version: entry.version });
  }
}

if (unknownProd.length) {
  console.warn('⚠ prod-зависимости без указанной лицензии (UNKNOWN):');
  for (const p of unknownProd) console.warn(`  - ${p.name}@${p.version}`);
}

if (violations.length || unknownProd.length) {
  console.error('\n✖ Гейт лицензий npm: обнаружены запрещённые/неизвестные лицензии');
  for (const v of violations) {
    console.error(`  - ${v.name}@${v.version}: ${v.license} [${v.kind}]${v.dev ? ' (dev)' : ''}`);
  }
  console.error(
    '\nЗапрещены copyleft-лицензии GPL/AGPL/LGPL/SSPL; перечень компонентов — THIRD-PARTY-NOTICES.md.',
  );
  process.exit(1);
}

console.log(`✓ Гейт лицензий npm: copyleft (GPL/AGPL/LGPL/SSPL) не обнаружен (${
  Object.keys(packages).length - 1
} пакетов проверено).`);

File: scripts/check_licenses.py
#!/usr/bin/env python3
"""Гейт лицензий Python-зависимостей (B52 — контроль copyleft).

Проходит по установленным дистрибутивам текущего окружения и завершается с
ошибкой (exit 1), если встречает strong-copyleft (GPL / AGPL / SSPL).

LGPL разрешён явно: драйвер PostgreSQL ``psycopg`` под LGPL-3.0 используется как
библиотека через Python-import (динамическое связывание), copyleft на наш код не
распространяется при использовании как библиотеки. Пакеты без данных о лицензии
выводятся как предупреждение, но сборку не роняют (у известных пакетов метаданные
всегда есть; жёсткий фейл по UNKNOWN дал бы ложные срабатывания на PyPI).

Без внешних зависимостей — только стандартный importlib.metadata.
"""

from __future__ import annotations

import sys
from importlib import metadata


def license_text(dist: metadata.Distribution) -> str:
    """Собирает все сведения о лицензии дистрибутива в одну строку."""
    meta = dist.metadata
    parts: list[str] = []
    for key in ("License-Expression", "License"):
        val = meta.get(key)
        if val and val != "UNKNOWN":
            parts.append(val)
    parts += [c for c in meta.get_all("Classifier", []) if c.startswith("License")]
    return " | ".join(parts)


def forbidden(text: str) -> str | None:
    """Вид запрещённой лицензии или None. LGPL/Lesser считаем разрешённым."""
    t = text.upper()
    if "AGPL" in t or "AFFERO" in t:
        return "AGPL"
    lesser = "LGPL" in t or "LESSER GENERAL PUBLIC" in t
    if not lesser and ("GPL" in t or "GENERAL PUBLIC LICENSE" in t):
        return "GPL"
    if "SSPL" in t or "SERVER SIDE PUBLIC" in t:
        return "SSPL"
    return None


def main() -> int:
    violations: list[tuple[str, str, str, str]] = []
    unknown: list[tuple[str, str]] = []
    seen: set[str] = set()

    for dist in metadata.distributions():
        name = dist.metadata.get("Name", "?")
        if name in seen:
            continue
        seen.add(name)
        text = license_text(dist)
        bad = forbidden(text)
        if bad:
            violations.append((name, dist.version, bad, text))
        elif not text.strip():
            unknown.append((name, dist.version))

    if unknown:
        print("⚠ пакеты без указанной лицензии (UNKNOWN):", file=sys.stderr)
        for pkg, ver in sorted(unknown):
            print(f"  - {pkg}=={ver}", file=sys.stderr)

    if violations:
        print("\n✖ Гейт лицензий Python: обнаружен запрещённый copyleft", file=sys.stderr)
        for pkg, ver, kind, text in violations:
            print(f"  - {pkg}=={ver}: [{kind}] {text}", file=sys.stderr)
        print(
            "\nЗапрещены GPL/AGPL/SSPL; LGPL (psycopg) разрешён намеренно. "
            "Перечень компонентов — THIRD-PARTY-NOTICES.md.",
            file=sys.stderr,
        )
        return 1

    print(
        f"✓ Гейт лицензий Python: strong-copyleft (GPL/AGPL/SSPL) не обнаружен "
        f"({len(seen)} дистрибутивов проверено; LGPL разрешён)."
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
